allow more than one DFE (#1078)
popsim-consortium/stdpopsim: docs/tutorial.rst

Incorporating selection
=======================

> There two general ways to incorporate selection into a simulation.

Reviewer: ```suggestion
There are two general ways to incorporate selection into a simulation.
```

Reviewer: so i dont get from this text what are the two general ways. But maybe I just need to go over everything.

Reviewer 2: The two ways are selection on intervals and selection on a single mutation site. I agree we should spell this out more clearly, also maybe linking these things to actual processes people will use them for, namely background selection and sweeps.

> Currently, both only work using the SLiM engine.

Reviewer: ```suggestion
Currently, both ways only work using the SLiM engine.
```

We can specify a :class:`distribution of fitness effects <.DFE>`
for all new mutations
(as in `the first example <sec_tute_genome_wide_dfe>`_)
and optionally a subset of the genome on which these occur
(as in `the second example <sec_tute_selection_single_gene>`_),
which can even be obtained from an :class:`.Annotation`
(see `the third example <sec_tute_selection_annotation>`_),
Additionally, we add a single mutation under selection,
as for instance in a `selective sweep <sec_tute_selective_sweep>`_.

.. _sec_tute_genome_wide_dfe:

1. Simulating with a genome-wide DFE
------------------------------------

To add mutations under selection to a simulation,
we need to tell the contig where to put the mutations,
and what distribution of selection coefficients they will have.

> To do this, we

Reviewer: ```suggestion
To do this, we need to:
```

- choose a distribution of fitness effects (a :class:`.DFE`),
- choose an :class:`.Annotation`, and

> - add these to the :meth:`Contig <.Contig.add_DFE>`,

Reviewer 2: wait, but in this example we are not choosing an annotation, we are saying to add the DFE to the whole contig.

Author: ah good point

    with the Annotation saying which portions of the genome the DFE
    applies to.

In this example, we'll add the Kim et al. HomSap/Gamma_K17 DFE to the
Gutenkunst et al. HomSap/OutOfAfrica_3G09 model.
We can see the DFEs available for a species in the catalog,
and get one using the :meth:`.Species.get_dfe` method.

.. code-block:: python

    import numpy as np

    species = stdpopsim.get_species("HomSap")
    contig = species.get_contig("chr1", length_multiplier=0.001)

>     dfe = species.get_dfe("Gamma_K17")

Reviewer: ```suggestion
    dfe = species.get_dfe("Gamma_K17")
    print(dfe) # describes the specific configuration of the DFE 
```

        f"{len(mut_info) - num_neutral} nonneutral mutations."
    )

    # There are 323 neutral mutations, and 420 nonneutral mutations.

.. _sec_tute_selection_single_gene:

2. Simulating selection in a single gene
----------------------------------------

Next, we'll simulate a 10kb gene flanked by 10kb neutral regions,
by specifying a particular interval to apply the HomSap/Gamma_K17 DFE to.
Contigs come by default covered by a neutral DFE,

> so all we need to do is apply the DFE to the middle region.

Reviewer: ```suggestion
so all we need to do is apply the DFE to the middle region (e.g. coding region).
```

> so all we need to do is apply the DFE to the middle region.

Reviewer 2: Should we use more words to describe what it means to apply a DFE on top of the existing neutral one?

            f"and {d} deleterious mutations."
        )

    # From 0 to 1000: 50 neutral mutations and 0 deleterious mutations.
    # From 1000 to 2000: 12 neutral mutations and 20 deleterious mutations.
    # From 2000 to 3000: 45 neutral mutations and 0 deleterious mutations.

> This verifies that there are only deleterious mutations in the center bit.

Reviewer: ```suggestion
This verifies that there are only deleterious mutations in the center bit (as expected under the Gamma_K17).
```

Reviewer 2: I think my wording here was actually very confusing. What we actually mean is: "This verifies that deleterious mutations can only be found in the center bit, given we applied the Gamma_K17 DFE just to this region."

Right now it can read like only deleterious mutations happen in the center bit, I think.

.. _sec_tute_selection_annotation:

3. Simulating selection on exons
--------------------------------

The catalog also has a certain number of *annotations* available,
obtained from Ensembl.
For instance, for humans we have:

.. code-block:: python

    for a in species.annotations:
        print(f"{a.id}: {a.description}")

    # ensembl_havana_104_exons: Ensembl Havana exon annotations on GRCh38
    # ensembl_havana_104_CDS: Ensembl Havana CDS annotations on GRCh38

> To simulate with the HomSap/Gamma_K17 DFE, applied this time

Reviewer 2: maybe `.., now applied to *all*...`

> to *all* exons on chromosome 21
> and remainder of the chromosome with have only neutral mutations),

Reviewer 3: ```suggestion
to *all* exons on chromosome 21
(the remainder of the chromosome will have only neutral mutations),
```

we extract the intervals from the :class:`.Annotation` object
and use this in :meth:`.Contig.add_DFE`:

.. code-block:: python

    species = stdpopsim.get_species("HomSap")
    dfe = species.get_dfe("Gamma_K17")
    contig = species.get_contig("chr20")
    model = species.get_demographic_model("OutOfAfrica_3G09")
    samples = model.get_samples(100, 100, 100)  # YRI, CEU, CHB

    exons = species.get_annotations("ensembl_havana_104_exons")

>     exon_intervals = exons.get_chromosome_annotations("chr20").astype("int")

Reviewer 4: i wonder if we should add a helper function here to apply the exonic intervals from a chrom / annotation. could save a line of code?

Author: https://github.com/popsim-consortium/stdpopsim/issues/1079#issuecomment-957110511

Reviewer 4: ah nice

>     exon_intervals = exons.get_chromosome_annotations("chr20").astype("int")

Reviewer 2: are we gonna change this and store everything as integer on the annotation side?

Author: yes! well, I think we should: https://github.com/popsim-consortium/stdpopsim/issues/1079

    contig.add_DFE(intervals=exon_intervals, DFE=dfe)

    engine = stdpopsim.get_engine("slim")
    ts = engine.simulate(
        model,
        contig,
        samples,
        seed=236,
        slim_scaling_factor=100,
        slim_burn_in=10,
    )

Note the very large scaling factor ($Q=100$) that we've used here to get this
to run fast enough to be used for a quick example!
This is *not* expected to be a good example because of this extreme scaling,
but nonetheless there is lower diversity in exons than outside of them:

.. code-block:: python

>     breaks, labels = contig.dfe_breakpoints()

Author: Note how the `dfe_breakpoints()` method gives us just what we need to pass to the tree sequence statistics.

Reviewer 2: aha! nice!

.. code-block:: python

> 

Reviewer 2: warning this API for sweeps isn't final? also do we have an issue to keep track of this?

Reviewer: I think so: #1082 #1085 #1053
